torchmultimodal/models/gpt.py

```diff
@@ -0,0 +1,418 @@
+        attn_probs = None
+        past_key_values = None
+
+        # Add head dim to attention mask for broadcasting
+        if attn_mask is not None:
+            attn_mask = get_extended_attention_mask(attn_mask)
+
+        attn_out = self._attn(
+            self.norm_attn(x),
+            attn_mask,
+            head_mask,
+            return_attn_weights,
+            use_cache=use_cache,
+            causal=causal,
+        )
+
+        if return_attn_weights:
+            x, attn_probs = attn_out
+        else:
+            x = attn_out
+
+        if use_cache:
+            past_key_values = self.attention.cache
+
+        x = x + self.dropout_attn(x)
+        x = self._mlp_block(self.norm_mlp(x))
+        x = x + self.dropout_mlp(x)
+
+        return TransformerLayerOutput(
+            hidden_states=x,
+            attention_weights=attn_probs,
+            past_key_values=past_key_values,
+        )
+
+    @checkpoint_wrapper
+    def _attn(
+        self,
+        x: Tensor,
+        attn_mask: Tensor,
+        head_mask: Tensor,
+        return_attn_weights: bool,
+        use_cache: bool,
+        causal: bool,
+    ) -> Union[Tensor, Tuple[Tensor, Tensor]]:
+        return self.attention(
+            x,
+            attention_mask=attn_mask,
+            head_mask=head_mask,
+            return_attn_weights=return_attn_weights,
+            use_cache=use_cache,
+            causal=causal,
+        )
+
+    @checkpoint_wrapper
+    def _mlp_block(self, x: Tensor) -> Tensor:
+        return self.mlp(x)
+
+
+class RightShift(nn.Module):
+    """Shift the input sequence by 1 unit to the right and prepend with start of sentence token.
+
+    Since the decoder progresses by taking the token it generates in the previous step, before it
+    has generated anything it needs a token to start with. Hence, the start-of-sentence (SOS) token.
+    The SOS token is a learnable parameter of the decoder and the choice of its initialization is taken
+    from VideoGPT: https://github.com/wilson1yan/VideoGPT/blob/master/videogpt/attention.py#L517
+    """
+
+    def __init__(self, embedding_dim):
+        super().__init__()
+        self.embedding_dim = embedding_dim
+        self.sos = nn.Parameter(
+            torch.FloatTensor(embedding_dim).normal_(std=0.02), requires_grad=True
+        )
+
+    def forward(self, x):
+        x_shape = list(x.shape)
+        x = x.flatten(start_dim=1, end_dim=-2)  # (batch, seq_len, embedding_dim)
+        sos = (
+            torch.ones(x_shape[0], 1, self.embedding_dim, dtype=torch.float32).to(
+                self.sos
+            )
+            * self.sos
+        )  # (batch, 1, embedding_dim)
+        sos = sos.type_as(x)
+        # Shift one unit to the right along dim ``seq_len``
+        x = torch.cat([sos, x[:, :-1, :]], axis=1)  # (batch, seq_len, embedding_dim)
+        x = x.view(*x_shape)
+        return x
+
+
+def _get_clones(module: nn.Module, n: int) -> nn.Module:
```

Comment:
should we make this a common utility since MDETR also uses this?

Reply:
I can move that to utils. @ebsmothers